web-common/src/features/dashboards/stores/validate-and-clean-explore-state.ts

```diff
@@ -0,0 +1,206 @@
+import type {
+  MetricsViewSpecDimension,
+  MetricsViewSpecMeasure,
+  V1ExploreSpec,
+  V1MetricsViewSpec,
+} from "@rilldata/web-common/runtime-client";
+
+/**
+ * Validates various fields in explore state.
+ * Removes any invalid state. Cascading merge should fill in remaining state.
+ *
+ * Note: Since this method will keep the state as is most of the time,
+ * it is more performant to do an in place modification of exploreState, rather than creating a copy.
+ * TODO: Look into doing this and cascading merge in a single place for performance when we have isolated states like FilterState etc
+ *
+ * Currently, it only acts on only a small section of the state.
+ *
+ * TODO: move all validations from convertUrlParamsToPreset and AdvancedMeasureCorrector here
+ */
+export function validateAndCleanExploreState(
+  metricsViewSpec: V1MetricsViewSpec,
+  exploreSpec: V1ExploreSpec,
+  exploreState: Partial<MetricsExplorerEntity>,
+) {
+  const errors: Error[] = [];
+
+  const measures = getMapFromArray(
+    metricsViewSpec.measures?.filter((m) =>
+      exploreSpec.measures?.includes(m.name!),
+    ) ?? [],
+    (m) => m.name!,
+  );
+  const dimensions = getMapFromArray(
+    metricsViewSpec.dimensions?.filter((d) =>
+      exploreSpec.dimensions?.includes(d.name!),
+    ) ?? [],
+    (d) => d.name!,
+  );
+
+  const errorsFromExploreView = validateAndCleanExploreViewState(
+    measures,
+    dimensions,
+    exploreSpec,
+    exploreState,
+  );
+  errors.push(...errorsFromExploreView);
+
+  return errors;
+}
+
+/**
+ * Looks at any invalid fields in explore view.
+ * Cleans up any invalid fields.
+ *
+ * Note: Since this method will keep the state as is most of the time,
+ * it is more performant to do an in place modification of exploreState, rather than creating a copy.
+ */
+function validateAndCleanExploreViewState(
+  measures: Map<string, MetricsViewSpecMeasure>,
+  dimensions: Map<string, MetricsViewSpecDimension>,
+  exploreSpec: V1ExploreSpec,
+  exploreState: Partial<MetricsExplorerEntity>,
+) {
+  const errors: Error[] = [];
+
+  if (exploreState.visibleDimensions) {
+    const selectedDimensions = exploreState.visibleDimensions.filter((d) =>
+      dimensions.has(d),
+    );
+    const missingDimensions = getMissingValues(
+      selectedDimensions,
+      exploreState.visibleDimensions,
+    );
+    if (missingDimensions.length) {
+      errors.push(getMultiFieldError("dimension", missingDimensions));
+    }
+
+    if (selectedDimensions.length > 0) {
+      // If there are any remaining valid dimensions then set it.
+      exploreState.allDimensionsVisible =
+        selectedDimensions.length === exploreSpec.dimensions?.length;
+      exploreState.visibleDimensions = selectedDimensions;
+    } else {
+      // Else remove the relevant fields so that cascading merge can set fields from other sources.
+      delete exploreState.allDimensionsVisible;
+      delete exploreState.visibleDimensions;
```

Comment:
This comment makes it sound like this function is tightly coupled to the cascading merge function.

Reply:
It is, I have a TODO to do them together. Out of scope of this PR